ng update errors have no line numbers

Currently an error during automatic migration looks like this:

>     ** Executing migrations for package '@ngrx/store' **
>     Cannot read property 'filter' of undefined

Not very helpful.  It should be more like this error message from @angular/core:

>     ** Executing migrations for package '@angular/core' **
>             ------ Static Query Migration ------
>             With Angular version 8, developers need to
>             explicitly specify the timing of ViewChild and
>             ContentChild queries. Read more about this here:
>             https://v8.angular.io/guide/static-query-migration
>             
>             The template migration strategy uses the Angular compiler internally and therefore projects that no longer build successfully after the update cannot use the template migration strategy. Please ensure there are no AOT compilation errors.
>             
>             Could not setup migration strategy for "src/tsconfig.app.json". The following error has been reported:
>             
>             Error: error TS100: Error during template compile of 'AppModule'
>               Function expressions are not supported in decorators in 'ɵ0'
>                 'ɵ0' contains the error at src/app/app.module.ts(29,21)
>                   Consider changing the function expression into an exported function.
>             
>             
>             Migration can be rerun with: "ng update @angular/core --from 7 --to 8 --migrate-only"
>             
>             ------------------------------------------------
>

--- a/modules/store/migrations/8_0_0-beta/index.spec.ts
+++ b/modules/store/migrations/8_0_0-beta/index.spec.ts
@@ -110,4 +110,48 @@ describe('Store Migration 8_0_0 beta', () => {
 
     expect(file).toBe(expected);
   });
+
+  it(`should not run schematics when not using named imports`, () => {
+    const contents = `
+      import * as store from '@ngrx/store';
+
+      @NgModule({
+        imports: [
+          CommonModule,
+          BrowserModule,
+          BrowserAnimationsModule,
+          HttpClientModule,
+          AuthModule,
+          AppRoutingModule,
+          store.StoreModule.forRoot(reducers),
+        ],
+        providers: [
+          {
+            provide: store.META_REDUCERS,
+            useValue: [fooReducer, barReducer]
+          }
+        ]
+        bootstrap: [AppComponent],
+      })
+      export class AppModule {}
+    `;
+
+    appTree.create('./app.module.ts', contents);
+    const runner = new SchematicTestRunner('schematics', collectionPath);
+
+    let logs: string[] = [];
+    runner.logger.subscribe(log => logs.push(log.message));
+
+    const newTree = runner.runSchematic(
+      `ngrx-${pkgName}-migration-02`,
+      {},
+      appTree
+    );
+    const file = newTree.readContent('app.module.ts');
+
+    expect(file).toBe(contents);
+
+    expect(logs.length).toBe(1);
+    expect(logs[0]).toMatch(/NgRx 8 Migration: Unable to run the schematics/);
+  });
 });

--- a/modules/store/migrations/8_0_0-beta/index.ts
+++ b/modules/store/migrations/8_0_0-beta/index.ts
@@ -1,5 +1,11 @@
 import * as ts from 'typescript';
-import { Rule, chain, Tree } from '@angular-devkit/schematics';
+import { tags, logging } from '@angular-devkit/core';
+import {
+  Rule,
+  chain,
+  Tree,
+  SchematicContext,
+} from '@angular-devkit/schematics';
 import {
   ReplaceChange,
   createReplaceChange,
@@ -10,7 +16,7 @@ import {
 const META_REDUCERS = 'META_REDUCERS';
 
 function updateMetaReducersToken(): Rule {
-  return (tree: Tree) => {
+  return (tree: Tree, context: SchematicContext) => {
     visitTSSourceFiles(tree, sourceFile => {
       const createChange = (node: ts.Node) =>
         createReplaceChange(
@@ -22,7 +28,11 @@ function updateMetaReducersToken(): Rule {
 
       const changes: ReplaceChange[] = [];
       changes.push(
-        ...findMetaReducersImportStatements(sourceFile, createChange)
+        ...findMetaReducersImportStatements(
+          sourceFile,
+          createChange,
+          context.logger
+        )
       );
       changes.push(...findMetaReducersAssignment(sourceFile, createChange));
 
@@ -37,11 +47,22 @@ export default function(): Rule {
 
 function findMetaReducersImportStatements(
   sourceFile: ts.SourceFile,
-  createChange: (node: ts.Node) => ReplaceChange
+  createChange: (node: ts.Node) => ReplaceChange,
+  logger: logging.LoggerApi
 ) {
+  let canRunSchematics = false;
+
   const metaReducerImports = sourceFile.statements
     .filter(ts.isImportDeclaration)
     .filter(isNgRxStoreImport)
+    .filter(p => {
+      canRunSchematics = Boolean(
+        p.importClause &&
+          p.importClause.namedBindings &&
+          (p.importClause!.namedBindings! as ts.NamedImports).elements
+      );
+      return canRunSchematics;
+    })
     .map(p =>
       (p.importClause!.namedBindings! as ts.NamedImports).elements.filter(
         isMetaReducersImportSpecifier
@@ -50,6 +71,15 @@ function findMetaReducersImportStatements(
     .reduce((imports, curr) => imports.concat(curr), []);
 
   const changes = metaReducerImports.map(createChange);
+  if (!canRunSchematics && changes.length === 0) {
+    logger.info(tags.stripIndent`
+      NgRx 8 Migration: Unable to run the schematics to rename \`META_REDUCERS\` to \`USER_PROVIDED_META_REDUCERS\`
+      in file '${sourceFile.fileName}'.
+
+      For more info see https://ngrx.io/guide/migration/v8#meta_reducers-token.
+    `);
+  }
+
   return changes;
 
   function isNgRxStoreImport(importDeclaration: ts.ImportDeclaration) {

